odl/operator/discr_ops.py

```diff
@@ -0,0 +1,446 @@
+
+    if edge_order is None:
+        edge_order = 2
+
+    dx = float(dx)
+
+    # create slice objects --- initially all are [:, :, ..., :]
+    # current slice
+    slice_out = [slice(None)] * ndim
+    # slices used to calculate finite differences
+    slice_node1 = [slice(None)] * ndim
+    slice_node2 = [slice(None)] * ndim
+    slice_node3 = [slice(None)] * ndim
+
+    # Numerical differentiation: 2nd order interior
+    slice_out[axis] = slice(1, -1)
+    slice_node1[axis] = slice(2, None)
+    slice_node2[axis] = slice(None, -2)
+    # 1D equivalent: out[1:-1] = (f[2:] - f[:-2])/2.0
+    np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+    out[slice_out] /= 2.0
+
+    # central differences
+    if zero_padding:
+        # Assume zeros for indices outside the domain of `f`
+
+        slice_out[axis] = 0
+        slice_node1[axis] = 1
+        # 1D equivalent: out[0] = (f[1] - 0)/2.0
+        out[slice_out] = f_data[slice_node1] / 2.0
+
+        slice_out[axis] = -1
+        slice_node2[axis] = -2
+        # 1D equivalent: out[-1] = (0 - f[-2])/2.0
+        out[slice_out] = - f_data[slice_node2] / 2.0
+
+    # one-sided differences
+    else:
+
+        # Numerical differentiation: 1st order edges
+        if f_data.shape[axis] == 2 or edge_order == 1:
+
+            slice_out[axis] = 0
+            slice_node1[axis] = 1
+            slice_node2[axis] = 0
+            # 1D equivalent: out[0] = (f[1] - f[0])
+            np.subtract(f_data[slice_node1], f_data[slice_node2],
+                        out[slice_out])
+
+            slice_out[axis] = -1
+            slice_node1[axis] = -1
+            slice_node2[axis] = -2
+            # 1D equivalent: out[-1] = (f[-1] - f[-2])
+            np.subtract(f_data[slice_node1], f_data[slice_node2],
+                        out[slice_out])
+
+        # Numerical differentiation: 2nd order edges
+        else:
+
+            slice_out[axis] = 0
+            slice_node1[axis] = 0
+            slice_node2[axis] = 1
+            slice_node3[axis] = 2
+            # 1D equivalent: out[0] = -(3*f[0] - 4*f[1] + f[2]) / 2.0
+            out[slice_out] = -(3.0 * f_data[slice_node1] - 4.0 * f_data[
+                slice_node2] + f_data[slice_node3]) / 2.0
+
+            slice_out[axis] = -1
+            slice_node1[axis] = -1
+            slice_node2[axis] = -2
+            slice_node3[axis] = -3
+            # 1D equivalent: out[-1] = (3*f[-1] - 4*f[-2] + f[-3]) / 2.0
+            out[slice_out] = (3.0 * f_data[slice_node1] - 4.0 * f_data[
+                slice_node2] + f_data[slice_node3]) / 2.0
+
+    # divide by step size
+    out /= dx
+
+    return out
+
+
+class DiscretePartDeriv(Operator):
+    """Calculate the discrete partial derivative along a given axis.
+
+        Calls helper function `finite_diff` to calculate finite difference.
+        Preserves the shape of the underlying grid.
+    """
+    # TODO: implement adjoint
+    # TODO: better name for `dx`
```

Comment:
I dont think `dx` is that bad, everyone understands what you mean instantly, which is the most important.